### PR TITLE
Jason Tip Calculator App 

### PR DESCRIPTION
**Pull request by jason-ng-zq99**
Tip Calculator App helps to calculate the amount you need to pay for the meal, inclusive of tips.

To do so, the calculator needs the following inputs: 
* Total amount of the meal
* Percentage of total meal cost to give as tip
* Number of people sharing the meal 

Constraints for the calculator: 
* Number of people cannot be zero (a warning will appear if this happens) 

Other notes:
* There are given buttons to choose the percentage for the tip, but you can also 
key in a custom percentage amount to calculate the interest

Desktop view:
Screenshot 2021-11-03 at 21 50 11

Mobile view:
Screenshot 2021-11-03 at 21 49 45

### CODE REVIEW

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  flex-direction: row;
+  justify-content: space-between;
+  height: 5vh;
+  line-height: 5vh;
+  background: hsl(189, 41%, 97%);
+  border-radius: 3px;
+  vertical-align: center;
+  ${(props) => props.peopleAmount === 0 && 'border: solid red 1px;'}
+  &:hover {
+    border: ${(props) => (props.peopleAmount === 0 ? 'border: solid red 1px;' : 'hsl(172, 67%, 45%) solid 1px')};
```
Comment (hui444) [inline src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: you used 'border:' in the true statement as well, remember to remove

File: src/store/tipCalculatorApp/action.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { Dispatch, GetState } from '../types'
+import { ActionTypes, CALCULATOR_ACTIONS } from './types'
+
+export const SetBillAmount = (newAmount: number) => async (dispatch: Dispatch<ActionTypes>) => {
```
Comment (hui444) [inline src/store/tipCalculatorApp/action.ts]: all these actions should be in camelCase!

File: src/store/tipCalculatorApp/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+type State = {
+  billAmount: number
+  peopleAmount: number
+  tipAmount: number
+  totalAmount: number
+}
+
+export const tipCalculatorApp: Reducer<State, ActionTypes> = (state = initialState, action) => {
+  switch (action.type) {
+    //this is to update the state if the action type is called
```
Comment (hui444) [inline src/store/tipCalculatorApp/reducer.ts]: remove comment

File: src/store/tipCalculatorApp/reducer.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { Reducer } from 'redux'
+import { ActionTypes, CALCULATOR_ACTIONS } from './types'
+
+//set an initial state for your variable
```
Comment (hui444) [inline src/store/tipCalculatorApp/reducer.ts]: remove comment

File: src/store/tipCalculatorApp/types.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/** Actions' types */
+export enum CALCULATOR_ACTIONS {
+  SET_BILL_AMOUNT = 'CALCULATOR_ACTIONS.SET_BILL_AMOUNT',
+  SET_TIP_AMOUNT = 'CALCULATOR_ACTIONS.SET_TIP_AMOUNT',
+  SET_PEOPLE_AMOUNT = 'CALCULATOR_ACTIONS.SET_PEOPLE_AMOUNT',
+  SET_TOTAL_AMOUNT = 'CALCULATOR_ACTIONS.SET_TOTAL_AMOUNT',
+  RESET = 'CALCULATOR_ACTIONS.RESET',
+}
+
+/** Actions */
```
Comment (hui444) [inline src/store/tipCalculatorApp/types.ts]: remove comments

**Review by hui444 — CHANGES_REQUESTED** [review]
minor issues, try and remove the comments unless you need them to explain something. and consider checking <= 0 instead of === 0 for the number of people; 

and might want to consider only showing error when user first enters the amount value

phone view a bit wonky tho,,,
Screenshot 2021-11-04 at 9 17 32 PM

otherwise, website looks good! do fix changes n should be ok!

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx (hunk 1 above)
Comment (jason-ng-zq99) [reply src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: Resolved in 9c54217

File: src/store/tipCalculatorApp/action.ts (hunk 2 above)
Comment (jason-ng-zq99) [reply src/store/tipCalculatorApp/action.ts]: Resolved in c0ae0a7

File: src/store/tipCalculatorApp/reducer.ts (hunk 3 above)
Comment (jason-ng-zq99) [reply src/store/tipCalculatorApp/reducer.ts]: Resolved in 8ca6ef0

File: src/store/tipCalculatorApp/reducer.ts (hunk 4 above)
Comment (jason-ng-zq99) [reply src/store/tipCalculatorApp/reducer.ts]: Resolved in 8ca6ef0

File: src/store/tipCalculatorApp/types.ts (hunk 5 above)
Comment (jason-ng-zq99) [reply src/store/tipCalculatorApp/types.ts]: Resolved in 8ca6ef0

File: src/store/tipCalculatorApp/action.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+export const setTotalAmount = () => async (dispatch: Dispatch<ActionTypes>, getState: GetState) => {
+  const { billAmount, peopleAmount, tipAmount } = getState().tipCalculatorApp
+  const newTotalAmount = billAmount / peopleAmount + tipAmount
+  dispatch({
+    type: CALCULATOR_ACTIONS.SET_TOTAL_AMOUNT,
+    totalAmount: newTotalAmount,
+  })
+}
+
+export const Reset = () => async (dispatch: Dispatch<ActionTypes>) => {
```
Comment (hui444) [inline src/store/tipCalculatorApp/action.ts]: are all the asyncs necessary?

File: src/pages/FrontendMentor/TipCalculatorApp/index.tsx [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                </TipGridItem>
+                <TipGridItem
+                  onClick={() => {
+                    dispatch(setBillAmount(newBillAmount))
+                    dispatch(setPeopleAmount(newPeopleAmount))
+                    dispatch(setTipAmount(50))
+                  }}
+                >
+                  50%
+                </TipGridItem>
```
Comment (hui444) [inline src/pages/FrontendMentor/TipCalculatorApp/index.tsx:139]: these look q similar, can consider making an object and map them to reduce duplicated code; but works too

File: src/store/tipCalculatorApp/action.ts (hunk 6 above)
Comment (jason-ng-zq99) [reply src/store/tipCalculatorApp/action.ts]: Resolved in 739bc12

File: src/store/JasonYbReduxShoppingList/actions.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { Dispatch, GetState } from '../types'
 import { ActionTypes, SHOPPING_LIST_ACTIONS, ShoppingItem } from './types'
 
 export const AddItemQuantity =
```
Comment (hui444) [inline src/store/JasonYbReduxShoppingList/actions.ts]: o my.. all these shouldve been camelCase

**Review by hui444 — COMMENTED** [review]
the actions in shoppinglist gotta change to camelCase....

also,, why is there NaN n consider making width bigger on phone, not sure why width is so small in mobile view..
Screenshot 2021-11-05 at 10 41 58 PM

otherwise code looks good

File: src/store/JasonYbReduxShoppingList/actions.ts (hunk 8 above)
Comment (jason-ng-zq99) [reply src/store/JasonYbReduxShoppingList/actions.ts]: Resolved in 80182f1

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  flex-direction: row;
+  justify-content: space-between;
+  height: 5vh;
+  line-height: 5vh;
+  background: hsl(189, 41%, 97%);
+  border-radius: 3px;
+  vertical-align: center;
+  ${(props) => props.peopleAmount <= 0 && 'border: solid red 1px;'}
+  &:hover {
+    border: ${(props) => props.peopleAmount <= 0 && 'hsl(172, 67%, 45%) solid 1px'};
```
Comment (hui444) [inline src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: might want to include 'border' in the portion where props.peopleAmount <= 0 is true instead

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  }
+`
+
+export const TipGridBox = styled.div`
+  display: grid;
+  height: 10vh;
+  grid-template-columns: repeat(3, minmax(0, 1fr));
+  grid-template-rows: repeat(2, minmax(0, 1fr));
+  grid-column-gap: 10px;
+  grid-row-gap: 10px;
```
Comment (hui444) [inline src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: can be done w `grid-gap: 10px` too

**Review by hui444 — APPROVED** [review]
minor issues but other than that lgtm! fix, deconflict n merge

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx (hunk 10 above)
Comment (jason-ng-zq99) [reply src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: Resolved in 9743465

File: src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx (hunk 9 above)
Comment (jason-ng-zq99) [reply src/pages/FrontendMentor/TipCalculatorApp/styles/TipCalculatorAppStyles.tsx]: Resolved in 70e675c

File: src/pages/FrontendMentor/TipCalculatorApp/index.tsx (hunk 7 above)
Comment (jason-ng-zq99) [reply src/pages/FrontendMentor/TipCalculatorApp/index.tsx:139]: Will try it next time! 👍🏼

**Pull request merged**